### PR TITLE
parent override child when annotation/labels conflicts && one revision will apply once only in force mode && AC.status CRD updated

### PR DESCRIPTION
**Pull request by wonderflow**
1) AC will override its component/workload; 
2) AC will override its trait; 
3) workload will override child resources

image

---

4) apply once only will check component revision
5) refactor CRD status of AppConfig to make workload has Dependency and appliedGeneration

### CODE REVIEW

**Review by ryanzhang-oss — APPROVED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go [hunk 1]
```diff
@@ -624,6 +625,12 @@ func applyOnceOnly(ac *v1alpha2.ApplicationConfiguration, mode core.ApplyOnceOnl
 					w.Reference.Name == d.GetName() {
 					// the workload matches applied resource
 					createdBefore = true
+
+					// for workload, when revision enabled, only when revision changed that can trigger to create a new one
+					if dLabels[oam.LabelOAMResourceType] == oam.ResourceTypeWorkload && w.ComponentRevisionName == dLabels[oam.LabelAppComponentRevision] {
+						// the revision is not changed, so return an error to abort creating it
+						return &GenerationUnchanged{}
+					}
```
[the remark this answers is not in the document]
Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go:633]: Yes

File: pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/revision_enable_test.go [hunk 2]
```diff
@@ -504,3 +504,322 @@ var _ = Describe("Test Component Revision Enabled with custom component revision
...
+			return k8sClient.Get(ctx, appConfigKey, &appConfig)
+		}, time.Second, 300*time.Millisecond).Should(BeNil())
+		By("Reconcile for new revision")
+		reconcileRetry(reconciler, req)
+
+		By("Check new revision workload created successfully")
+		Eventually(func() error {
+			var workloadKey = client.ObjectKey{Namespace: namespace, Name: compName + "-v2"}
+			return k8sClient.Get(ctx, workloadKey, &wr)
+		}, time.Second, 300*time.Millisecond).Should(BeNil())
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/revision_enable_test.go:791]: Move `reconcileRetry` into `Eventually` block can avoid unstable test result.

File: pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/revision_enable_test.go [hunk 3]
```diff
@@ -504,3 +504,322 @@ var _ = Describe("Test Component Revision Enabled with custom component revision
...
+		By("Reconcile")
+		reconciler.applyOnceOnlyMode = "force"
+		reconcileRetry(reconciler, req)
+
+		By("Check workload created successfully")
+		var workloadKey1 = client.ObjectKey{Namespace: namespace, Name: compName + "-v1"}
+		Eventually(func() error {
+
+			return k8sClient.Get(ctx, workloadKey1, &wr)
+		}, 3*time.Second, 300*time.Millisecond).Should(BeNil())
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/revision_enable_test.go:678]: Move `reconcileRetry` into `Eventually` block can avoid unstable test results.

Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/revision_enable_test.go]: good idea

**Pull request merged**